### PR TITLE
Updated needed for changes/updates in #760 of mxcubecore

### PR DESCRIPTION
**Pull request by marcus-oscarsson**
Updated needed for changes/updates in #760 of mxcubecore

**Updates done:**

- SSX related components (chip)
- Adapted to latest workflow code GPHL and ESRF for A1, adds the possibility to run custom scripts for phase change and centering on the MD 
- Some cleanup and minor fixes

### CODE REVIEW

**Comment by marcus-oscarsson** [comment]
Looks like there is some inconsistency with my local `mxcubecore` and the one on develop that fails a bunch of tests ill try to get that fixed ASAP.

File: mxcube3/core/adapter/diffractometer_adapter.py [hunk 1]
```diff
@@ -15,13 +15,13 @@ def __init__(self, ho, *args, **kwargs):
...
     def _state_change(self, *args, **kwargs):
-        self.state_change(**kwargs)
+        self.state_change(*args, **kwargs)
 
     def stop(self):
         pass
 
     def state(self):
-        return "READY" if self._ho.is_ready() else "BUSY"
+        return "READY" if self._ho.get_state().name == "READY" else "BUSY"
```
Comment (rhfogh) [inline mxcube3/core/adapter/diffractometer_adapter.py]: What is the reason to bypass the is_ready function? Unless there is some object that does not have it and cannot be given it, it would be better to use always the same function for readiness testing.

File: mxcube3/core/components/queue.py [hunk 2]
```diff
@@ -1634,6 +1638,9 @@ def add_workflow(self, node_id, task):
         parent_entry.enqueue(group_entry)
         group_entry.enqueue(dc_entry)
 
+        if task["parameters"]["wfpath"] != "Gphl":
```
Comment (rhfogh) [inline mxcube3/core/components/queue.py]: Could this be done in the same lines (and at teh same place in the code) as for the equivalent Gphl workflow? Or do you know a reason why this must be done in two different places a few lines apart?

File: mxcube3/core/components/queue.py [hunk 3]
```diff
@@ -2313,6 +2332,11 @@ def get_default_task_parameters(self, task_name):
 
         schema = self.get_task_schema(data_model) if data_model else {}
 
+        try:
```
Comment (rhfogh) [inline mxcube3/core/components/queue.py:2333]: This looks like a system for controlling the UI rendering. Maybe I should adapt my GPhL UI to use this - but someone would have to explain it to me first.

File: mxcube3/routes/queue.py [hunk 4]
```diff
@@ -278,6 +278,20 @@ def get_avilable_tasks():
         resp.status_code = 200
         return resp
 
+    @bp.route("/update_dependent_field", methods=["POST"])
+    @server.restrict
+    def update_dependent_field():
+        """
+        Returns a list of all available tasks
```
Comment (rhfogh) [inline mxcube3/routes/queue.py]: Does this doc string need updating?

File: mxcube3/routes/samplecentring.py [hunk 5]
```diff
@@ -76,7 +76,7 @@ def set_image_size():
         params = request.get_json()
 
         res = app.sample_view.set_image_size(
-            int(params["width"]), int(params["height"])
+            float(params["width"]), float(params["height"])
```
Comment (rhfogh) [inline mxcube3/routes/samplecentring.py:79]: Is this set in floating point pixels/ or in mm? or what?

**Review by rhfogh — CHANGES_REQUESTED** [review]
Looks nice (though much of the javascript stuff goes over my head). I'd approve, except for the minor points raised.

File: mxcube3/core/components/lims.py [hunk 6]
```diff
@@ -322,7 +323,10 @@ def select_proposal(self, proposal):
             HWR.beamline.session.proposal_number = proposal_info.get("Proposal").get(
                 "number", ""
             )
-            todays_session = HWR.beamline.lims.get_todays_session(proposal_info)
+
+            todays_session = HWR.beamline.lims.get_todays_session(
```
Comment (meguiraun) [inline mxcube3/core/components/lims.py:327]: we will have troubles with this `create_session=False`, why did you change it?

File: mxcube3/core/components/queue.py [hunk 7]
```diff
@@ -1382,6 +1383,9 @@ def _create_queue_entry(self, task, task_name):
         Return:
             (tuple): (model, entry)
         """
+        if not task["parameters"]["osc_range"]:
```
Comment (meguiraun) [inline mxcube3/core/components/queue.py:1386]: I dont understand this 🧐...

`osc_range` comes with a `null` value? Should not this be fixed in the ui?

File: mxcube3/core/components/queue.py [hunk 8]
```diff
@@ -2294,6 +2301,18 @@ def add_centring(self, _id, params):
 
         return {"QueueId": new_node, "Type": "Centring", "Params": params}
 
+    def update_dependent_field(self, task_name, data):
```
Comment (meguiraun) [inline mxcube3/core/components/queue.py:2302]: what is this for? What is `task_name` here? `task_name` for me would be `DataCollection`...

File: mxcube3/core/components/samplechanger.py [hunk 9]
```diff
@@ -432,53 +433,66 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):
...
-
-                logging.getLogger("user_level_log").info("Centring ...")
-                centring_result = async_result.get()
-                if centring_result["valid"]:
-                    logging.getLogger("user_level_log").info("Centring done !")
-                else:
-                    if centring_method == queue_entry.CENTRING_METHOD.FULLY_AUTOMATIC:
-                        raise queue_entry.QueueSkippEntryException(
-                            "Could not center sample, skipping", ""
+        use_custom_centring_routine = HWR.beamline.diffractometer.get_property(
```
Comment (meguiraun) [inline mxcube3/core/components/samplechanger.py:437]: i like this... What are you using this for? (XRay centring comes to my mind)

File: mxcube3/core/components/sampleview.py [hunk 10]
```diff
@@ -560,10 +562,15 @@ def take_snapshots(self, snapshots=None, _do_take_snapshot=_do_take_snapshot):
                 hasattr(diffractometer_object, "set_phase")
                 and diffractometer_object.get_current_phase() != "Centring"
             ):
-                logging.getLogger("user_level_log").info(
-                    "Moving Diffractometer to CentringPhase"
+                use_custom_snapshot_routine = diffractometer_object.get_property(
```
Comment (meguiraun) [inline mxcube3/core/components/sampleview.py:565]: i wonder if we should keep this get_prop inside the dm object.

Also, would not this be HWR.beamline.diffractometer.....

edit: wait... is this actually in use?

File: mxcube3/core/components/sampleview.py [hunk 11]
```diff
@@ -560,10 +562,15 @@ def take_snapshots(self, snapshots=None, _do_take_snapshot=_do_take_snapshot):
...
+                    "custom_snapshot_script_dir", None
                 )
-                diffractometer_object.set_phase("Centring", wait=True, timeout=200)
+                if not use_custom_snapshot_routine:
+                    logging.getLogger("user_level_log").info(
+                        "Moving Diffractometer to CentringPhase Not done for tests (DN)"
```
Comment (meguiraun) [inline mxcube3/core/components/sampleview.py:570]: wrong msg

File: ui/src/components/Tasks/Characterisation.js [hunk 12]
```diff
@@ -448,16 +451,16 @@ Characterisation = connect((state) => {
       beam_size: state.sampleview.currentAperture,
       resolution: (state.taskForm.sampleIds.constructor !== Array
         ? state.taskForm.taskData.parameters.resolution
-        : state.beamline.hardwareObjects.resolution.value),
+        : toFixed(state, "resolution")),
```
Comment (meguiraun) [inline ui/src/components/Tasks/Characterisation.js:454]: I dont understand this change, from where the values comes now?

File: mxcube3/core/components/queue.py (hunk 7 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/queue.py:1386]: Yes, you are indeed right that it seems a bit odd. However now a days and especially with SSX oscillation range might not be set or used at all hence its `Null`. We could perhaps set it to 0 instead. There is some thought work to do and its to remove the "composite collection" dictionary that we are using for the "traditional" oscillation experiments.

Comment (marcus-oscarsson) [reply mxcube3/core/components/queue.py:1386]: I think in the end it just comes down to actually doing this as we have been speaking about it for sometime. I think we should discuss as part of our roadmap discussion because I think it also should be treated together with the queue discussion.

File: mxcube3/core/components/queue.py (hunk 2 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/queue.py]: yes, I think two different people did a very similar change and then we did not pick this up when we merged.

File: mxcube3/core/components/queue.py (hunk 8 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/queue.py:2302]: :), I guess this one is fine as we spoke about this during our meeting today,

File: mxcube3/core/components/queue.py (hunk 3 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/queue.py:2333]: Yes it probably could, we should as we said earlier have a look at if what we have today satisfy your requirements.

File: mxcube3/core/components/samplechanger.py (hunk 9 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/samplechanger.py:437]: Well, its mostly used on A1 where we wrote some PMAC routine to speed up phase change and perform cantering on the MD computer. Its work mostly done by **Didier Nurizzo**. There are actually also some workarounds to issues of the MD firmware, so we can now specify weather to use those or not

File: mxcube3/core/components/sampleview.py (hunk 10 above)
Comment (marcus-oscarsson) [reply mxcube3/core/components/sampleview.py:565]: Yes it is unfortunately used and we should really really move away from it. Ill give it a try but it needs to be tested a bit before I remove it. It was used to get the overlays from the UI but we are not taking snapshots that way any longer. I will try to make a PR, shortly, removing the entire enable_snapshot "patch" and add something cleaner

File: mxcube3/routes/queue.py (hunk 4 above)
Comment (marcus-oscarsson) [reply mxcube3/routes/queue.py]: Oh yes

File: mxcube3/routes/samplecentring.py (hunk 5 above)
Comment (marcus-oscarsson) [reply mxcube3/routes/samplecentring.py:79]: Its microscope (OAV) image size so its in pixels

File: ui/src/components/Tasks/Characterisation.js (hunk 12 above)
Comment (marcus-oscarsson) [reply ui/src/components/Tasks/Characterisation.js:454]: Ah, its the application of the values specified in the [ui.yaml ](https://github.com/mxcube/mxcubeweb/blob/develop/test/HardwareObjectsMockup.xml/mxcube-web/ui.yaml) file. Basically just making sure that the precision in that file is used instead of like before when it was a bit fixed.

Comment (meguiraun) [reply ui/src/components/Tasks/Characterisation.js:454]: i meant the actual values written to the fields, I double checked and now I see the toFixed method, so all good

Comment (marcus-oscarsson) [reply ui/src/components/Tasks/Characterisation.js:454]: :) cool

**Review by meguiraun — APPROVED** [review]

File: mxcube3/routes/samplecentring.py (hunk 5 above)
Comment (rhfogh) [reply mxcube3/routes/samplecentring.py:79]: But not integer values, apparently?
